libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp

```
@@ -0,0 +1,1227 @@
+    // we need to update the file config, but that means finding the file config
+    for (int fd = 0; fd < MAX_OPEN_FILES; fd++) {
+        if (open_files[fd] != nullptr && open_files[fd]->filename == filename) {
+            open_files[fd]->mtime = mtime_sec;
+            return true;
+        }
+    }
+    return false;
+}
+
+int AP_Filesystem_FlashMemory_LittleFS::unlink(const char *pathname)
+{
+    FS_CHECK_ALLOWED(-1);
+    WITH_SEMAPHORE(fs_sem);
+    ENSURE_MOUNTED();
+    LFS_CHECK(lfs_remove(&fs, pathname));
+    return 0;
+}
+
+int AP_Filesystem_FlashMemory_LittleFS::mkdir(const char *pathname)
+{
+    FS_CHECK_ALLOWED(-1);
+    WITH_SEMAPHORE(fs_sem);
+    ENSURE_MOUNTED();
+    LFS_CHECK(lfs_mkdir(&fs, pathname));
+    return 0;
+}
+
+typedef struct {
+    lfs_dir_t dir;
+    struct dirent entry;
+} lfs_dir_entry_pair;
+
+void *AP_Filesystem_FlashMemory_LittleFS::opendir(const char *pathdir)
+{
+    int retval;
+
+    FS_CHECK_ALLOWED(nullptr);
+    WITH_SEMAPHORE(fs_sem);
+    ENSURE_MOUNTED_NULL();
+
+    lfs_dir_entry_pair *result = new lfs_dir_entry_pair;
+    if (!result) {
+        errno = ENOMEM;
+        return nullptr;
+    }
+
+    retval = lfs_dir_open(&fs, &result->dir, pathdir);
+    if (retval < 0) {
+        delete result;
+        errno = errno_from_lfs_error(retval);
+        return nullptr;
+    }
+
+    memset(&result->entry, 0, sizeof(result->entry));
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX || CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    result->entry.d_reclen = sizeof(result->entry);
+#endif
+
+    return result;
+}
+
+struct dirent *AP_Filesystem_FlashMemory_LittleFS::readdir(void *ptr)
+{
+    FS_CHECK_ALLOWED(nullptr);
+    WITH_SEMAPHORE(fs_sem);
+
+    lfs_info info;
+    lfs_dir_entry_pair *pair = static_cast<lfs_dir_entry_pair*>(ptr);
+    if (!pair) {
+        errno = EINVAL;
+        return nullptr;
+    }
+
+    if (!lfs_dir_read(&fs, &pair->dir, &info)) {
```

Needs an error check. For some strange reason this tries to return `LFS_ERR_CORRUPT` when using the clear log button in QGroundControl but it gets stuck forever since the error gets ignored here. Needs further investigation.

---

Good catch. Fixed